src/hypothesis/searchstrategy/types.py

```diff
@@ -194,12 +196,20 @@ def really_inner(thing):
     def resolve_Type(thing):
         if thing.__args__ is None:
             return st.just(type)
-        inner = thing.__args__[0]
-        if getattr(inner, '__origin__', None) is typing.Union:
-            return st.sampled_from(inner.__args__)
-        elif hasattr(inner, '__union_params__'):  # pragma: no cover
-            return st.sampled_from(inner.__union_params__)
-        return st.just(inner)
+        args = (thing.__args__[0],)
+        if getattr(args[0], '__origin__', None) is typing.Union:
+            args = args[0].__args__
+        elif hasattr(args[0], '__union_params__'):  # pragma: no cover
+            args = args[0].__union_params__
+        if isinstance(typing._ForwardRef, type):  # pragma: no cover
+            # Duplicate check from from_type here - only paying when needed.
+            for a in args:
+                if type(a) == typing._ForwardRef:
+                    raise ResolutionFailed(
```

Comment:
Nit: It would be nice having this factored out into some sort of `check_not_forward_ref` helper method I think - we've got the same error message in two places, which is otherwise easy to get out of sync.